README.md
# Flask-Allow 
Flask-Allow is an extension for Flask that adds support for white and/or black listing 
IP addresses and/or IP networks and provide an access log to your application.

# Why this extension
Whenever the web application runs behind a reverse proxy that is located on a different system
in the network, you want to grant access to the proxy but exclude all other hosts in the network.

    ALLOW:  proxy-host.your.domain.tld
    ALLOW:  localhost
    DENY:   0.0.0.0/0

These rules allow the host **proxy-host.your.domain.tld** and **localhost** to pass in to the web
application. **localhost** is there to give administrators access to the web application when 
running on the same machine. The **0.0.0.0/0** blocks all other addresses.

Not the order in which you configure the rules is important. for example swaping **localhost** and
**0.0.0.0/0** shall exclude the **localhost**.

Why should you use a reverse proxy, read all about it in the 
[Link](https://flask.palletsprojects.com/en/2.3.x/deploying/nginx/)


# Version
Currently this supports and is tested with Flask 2.x.x. therefore the version of this package
is version 2.0.x. Tested with Python version 3.8, 3.9, 3.10 and 3.11. 


# Licence
Flask-Allow is licenced under GPL-2.0-only, see the LICENCE.md for more information.


# Installing
Install and update using pip.

```bash
    $ pip install -U Flask-Allow
```


# Configuration
This extension has two configuration items:
* **ADDRESS_RESTRICTION**
* **ACCESS_LOG**

The attribute **ADDRESS_RESTRICTION** is a list of dictionaries with one or two items
* **ALLOW**; the IP address or network address to be granted access. 
* **DENY**; the IP address or network address to be denied access.

For IP network addresses it must be coded as <IP-address>/<no-bits>, for example:

    172.16.0.0/16

For allowing or denying single hosts you may even write the fqdn of the host you want to exclude;

    DENY: test.example.com
    ALLOW: prod.example.com

The attribute **ACCESS_LOG** may be a filename or a dictionary, it uses rotating file logger. 
When using a dictionary the following items may be provided;
* **filename**; sets the filename for the access log. 
* **maxBytes**; sets the maximum size of the log file, default is 5242880.
* **backupCount**; sets the maximum historical log files kept, default is 7. 
* **formatter**; Sets the log file formatter, default is "%(asctime)s - %(levelname)7s - %(message)s" 

The logger created is called **flask.allow**, when configured the log level is set the INFO.


# Simple example
The following example sets up a web server on host address 0.0.0.0 (all networks) with port 5000.
An access log is created and only the localhost address is allowed to enter the application, all 
other addresses receive a HTTP 403 error.

```python
import flask
from flask_allow import FlaskAllow

app = flask.Flask( __name__ )
app.config[ 'ACCESS_LOG' ]  = "access.log"
app.config[ 'ADDRESS_RESTRICTION' ] = [
    {
        "ALLOW":    "127.0.0.1",             # Allow localhost
        "DENY":     "0.0.0.0/0"              # Deny the rest
    }
]
FlaskAllow( app )

@app.route('/')
def index():
    return "Hello world", 200

app.run( '0.0.0.0', 5000 )
    
```
**NOTE:** The class FlaskAllow should be initialized before any @before_request decorators
are being called, this to ensure that Flask-Allow is the first to check in incomming request.

The following log output is from the test_flask_allow.py script. 
```log
2023-12-03 07:34:27,883 -    INFO - Access log started
2023-12-03 07:34:28,886 -    INFO - 127.0.0.1 allowed by rule 127.0.0.1/32 http://localhost:5000/ 
2023-12-03 07:34:28,893 -    INFO - 127.0.0.1 allowed by rule 127.0.0.1/32 http://localhost:5000/ python-requests/2.31.0
2023-12-03 07:34:28,903 -   ERROR - 192.168.110.2 denied by rule 0.0.0.0/0 http://matrix:5000/ python-requests/2.31.0
```


# Contributing
For guidance on setting up a development environment and how to make a contribution to 
flask-access, see the contributing guidelines.


# Donate
The Pallets organization develops and supports Flask and other popular packages. 
In order to grow the community of contributors and users, and allow the maintainers to devote 
more time to the projects, [please donate today](https://palletsprojects.com/donate)


# Links
* [Changes](https://github.com/pe2mbs/flask-allow/CHANGED.md)
* [PyPI Releases](https://pypi.org/project/flask_allow/)
* [Source Code](https://github.com/pe2mbs/Flask-Allow)
* [Issue Tracker](https://github.com/pe2mbs/Flask-Allow/issues)
* [Website](https://github.com/pe2mbs/Flask-Allow)


CHANGES.md
# Version 2.0.0 - 3 December 2023
Release version for Flask 2.x.x framework. 

tests/test_flask_allow.py
import pytest
import requests
import ipaddress
import socket
from flask_allow import FlaskAllow
from tests.flask_test_server import flask_test_server


@pytest.fixture(scope="session")
def startFlask():
    import threading
    import time
    from http.client import HTTPConnection
    t = threading.Thread( target = flask_test_server )
    t.daemon = True
    t.start()
    time.sleep(1)
    retries = 5
    response = None
    while retries > 0:
        conn = HTTPConnection("localhost:5000")
        try:
            conn.request("HEAD", "/")
            response = conn.getresponse()
            if response is not None:
                yield t
                break

        except ConnectionRefusedError:
            time.sleep(60)
            retries -= 1

    if retries == 0:
        raise RuntimeError("Failed to start http server")

    return


def test_validate_address():
    assert( FlaskAllow.networkMask( '192.168.110.123' ) )
    assert( FlaskAllow.networkMask( '2001:0db8:85a3:0000:0000:8a2e:0370:7334' ) )
    assert( FlaskAllow.networkMask( '192.168.110.0' ) )
    assert( FlaskAllow.networkMask( 'example.org' ) )
    assert( FlaskAllow.networkMask( 'localhost' ) )
    assert( FlaskAllow.networkMask( "2001:db00::0/24" ) )
    assert( FlaskAllow.networkMask( '192.168.110.0/24' ) )
    with pytest.raises(ipaddress.AddressValueError):
        assert( FlaskAllow.networkMask( 'notexists.example.org' ) )

    return


def test_access_localhost( startFlask ):
    r = requests.get( 'http://localhost:5000' )
    assert( r.status_code == 200 )
    return


def test_access_publichost( startFlask ):
    host = socket.gethostname()
    r = requests.get( f'http://{host}:5000' )
    assert( r.status_code == 403 )
    return


tests/flask_test_server.py
#!/usr/bin/python3
# -*- coding: utf-8 -*-
"""
    This is a simple Flask web server with flask_allow for testing purposes.
    This is being called from the pytest script test_flask_allow.py

"""
import flask
import traceback
import logging
import click
from flask_allow import FlaskAllow


def flask_test_server( verbose = False ):
    # Force logging of the webserver off
    if not verbose:
        log = logging.getLogger('werkzeug')
        log.setLevel(logging.ERROR)
        log.disabled = True

    try:
        app = flask.Flask( __name__ )
        # Force logging of the webserver off
        if not verbose:
            app.logger.disabled = True

        app.config[ 'TESTING' ]     = True
        app.config[ 'ACCESS_LOG' ]  = "access.log"
        app.config[ 'ADDRESS_RESTRICTION' ] = [
            {
                "ALLOW":    "127.0.0.1",             # Allow localhost
                "DENY":     "0.0.0.0/0"              # Deny the rest
            }
        ]

        # Force logging click off
        def secho(text, file=None, nl=None, err=None, color=None, **styles):
            pass

        def echo(text, file=None, nl=None, err=None, color=None, **styles):
            pass

        if not verbose:
            click.echo = echo
            click.secho = secho

        FlaskAllow( app )

        @app.route('/')
        def index():
            return "Hello world", 200

        app.run( '0.0.0.0', 5000 )

    except Exception:
        print( traceback.format_exc() )

    return


if __name__ == '__main__':
    flask_test_server( verbose = True )


flask_allow/__init__.py
#   flask_allow allows white/black listing of ip addresses/networks.
#   Copyright (C) 2023 Marc Bertens-Nguyen
#
#   This program is free software; you can redistribute it and/or
#   modify it under the terms of the GNU General Public License
#   as published by the Free Software Foundation; only version 2
#   of the License (GPL-2.0-only).
#
#   This program is distributed in the hope that it will be useful,
#   but WITHOUT ANY WARRANTY; without even the implied warranty of
#   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#   GNU General Public License for more details.
#
#   You should have received a copy of the GNU General Public License
#   along with this program; if not, write to the Free Software
#   Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
#
from flask_allow.allow import FlaskAllow    # noqa


flask_allow/exceptions.py
#   Flask-Allow allows white/black listing of ip addresses/networks.
#   Copyright (C) 2023 Marc Bertens-Nguyen
#
#   This program is free software; you can redistribute it and/or
#   modify it under the terms of the GNU General Public License
#   as published by the Free Software Foundation; only version 2
#   of the License (GPL-2.0-only).
#
#   This program is distributed in the hope that it will be useful,
#   but WITHOUT ANY WARRANTY; without even the implied warranty of
#   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#   GNU General Public License for more details.
#
#   You should have received a copy of the GNU General Public License
#   along with this program; if not, write to the Free Software
#   Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
#
__version__ = '2.0.0'
__author__  = 'Marc Bertens-Nguyen'
__licence__ = 'GPL-2.0-only'


class InvalidAccessLog( Exception ):
    def __init__( self, message, data ):
        super().__init__( message )
        self.data = data
        return


flask_allow/allow.py
#   Flask-Allow allows white/black listing of ip addresses/networks.
#   Copyright (C) 2023 Marc Bertens-Nguyen
#
#   This program is free software; you can redistribute it and/or
#   modify it under the terms of the GNU General Public License
#   as published by the Free Software Foundation; only version 2
#   of the License (GPL-2.0-only).
#
#   This program is distributed in the hope that it will be useful,
#   but WITHOUT ANY WARRANTY; without even the implied warranty of
#   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#   GNU General Public License for more details.
#
#   You should have received a copy of the GNU General Public License
#   along with this program; if not, write to the Free Software
#   Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
#
from typing import Union
import os
import ipaddress
import socket
import logging
import logging.handlers
from mako.template import Template
from flask import Flask, request, abort
from flask_allow.exceptions import InvalidAccessLog


__version__ = '2.0.0'
__author__  = 'Marc Bertens-Nguyen'
__licence__ = 'GPL-2.0-only'


class FlaskAllow( object ):
    """ This class is intended as extention to Flask to restrict the IP addresses
    comming into the web application and access log file

        app = Flask()
        restrict = FlaskAllow()
        restrict.init_app( app )

    or
        app = Flask()
        restrict = FlaskAllow( app )

    Configuration
    The ADDRESS_RESTRICTION is a list if dictionaries, "ALLOW", "DENY" or both are allowed,
    the ALLOW is is checked first.

        app.config[ "ADDRESS_RESTRICTION" ] = [
            {
                "ALLOW":    "192.168.1.0/24",       # Allow subnet 192.168.1.*
            },
            {
                "ALLOW":    "127.0.0.1"             # Allow localhost
                "DENY": "0.0.0.0/0",            # Allow subnet 192.168.1.*
            },
        ]

    or in YAML format
        ADDRESS_RESTRICTION:
        -   ALLOW:  192.168.1.0/24
        -   ALLOW:  127.0.0.1
            DENY:   0.0.0.0/0

    The access logger is using the RotatingFileHandler.
    For the access log just provide the log filename;

        app.config[ "ACCESS_LOG" ] = 'access.log'

    Or with more control;

        app.config[ "ACCESS_LOG" ] = {
            "filename": "access.log",
            "maxBytes": 123456789
            "backupCount": 7
            "formatter": "%(asctime)s - %(levelname)7s - %(message)s"
        }

    Or in YAML format

        ACCESS_LOG:
            filename: access.log
            maxBytes: 123456789
            backupCount: 7
            formatter: "%(asctime)s - %(levelname)7s - %(message)s"

    """
    def __init__( self, app: Flask = None ):
        """ Initialize the class, create an empty list (tuple), set up the logger

        :param app:     Optional the Flask application instance, when provided the init_app() function is directly called

        :raises:        ipaddress.AddressValueError, flask_allow.allow.InvalidAccessLog, ValueError

        :returns:       None
        """
        self.__allowdeny    = tuple()
        self.__app          = app
        self.__access       = logging.getLogger( "flask.allow" )
        self.__access.setLevel( logging.NOTSET )
        if app is not None:
            self.init_app( app )

        return

    ACCESS_LOG_FORMATTER    = "%(asctime)s - %(levelname)7s - %(message)s"
    ACCESS_LOG_HANDLER      = {
        'filename': None,
        'maxBytes': 5242880,
        'backupCount': 7
    }

    def create_access_log( self ):
        accessLog = self.__app.config.get( "ACCESS_LOG" )

        def update_handlder_data( filename, maxBytes = 5242880, backupCount = 7, **kwargs ):
            if '${' in filename:
                filename = Template( text = filename ).render( **dict( self.__app.config ) )
                if not os.path.isfile( filename ):
                    raise Exception( f"{filename} is not a valid filename" )

            self.ACCESS_LOG_HANDLER[ 'filename' ]       = filename
            self.ACCESS_LOG_HANDLER[ 'maxBytes' ]       = maxBytes
            self.ACCESS_LOG_HANDLER[ 'backupCount' ]    = backupCount
            return

        if accessLog is None:
            return

        elif isinstance( accessLog, dict ):
            formatter = accessLog.get( 'formatter', self.ACCESS_LOG_FORMATTER )
            update_handlder_data( **accessLog )

        elif isinstance( accessLog, str ):
            self.ACCESS_LOG_HANDLER[ 'filename' ]   = accessLog
            formatter = self.ACCESS_LOG_FORMATTER

        else:
            raise InvalidAccessLog( "ACCESS_LOG should be a filename or a dictionary", accessLog )

        if isinstance( self.ACCESS_LOG_HANDLER[ 'filename' ], str ):
            handler = logging.handlers.RotatingFileHandler( **self.ACCESS_LOG_HANDLER )
            handler.setFormatter( logging.Formatter( formatter ) )
            self.__access .setLevel( logging.INFO )
            self.__access .addHandler( handler )
            self.__access .info( "Access log started" )
            if len( self.__allowdeny ) == 0:
                self.__app.before_request_funcs.setdefault( None, [] ).append( self._access_log )

        return

    def init_app( self, app: Flask ) -> None:
        """ Initialize the class

        :param app:     Optional the Flask application instance, when provided the init_app() function is directly called

        :raises:        ipaddress.AddressValueError

        :returns:       None

        When the ALLOW is set to IP addrsses or neyworks, the 0.0.0.0/0 shall be added to the deny list

        """
        self.__app = app
        restrict = app.config.get( "ADDRESS_RESTRICTION" )
        if not isinstance( restrict, list ):
            app.logger.warning( "ADDRESS_RESTRICTION not configured" )
            return

        allowdenyList = []
        for allowdeny in restrict:
            if isinstance( allowdeny, dict ):
                a = allowdeny.get( "ALLOW" )
                if isinstance( a, str ):
                    allowdenyList.append( ( "A", self.networkMask( a ) ) )

                a = allowdeny.get( "DENY" )
                if isinstance( a, str ):
                    allowdenyList.append( ( "D", self.networkMask( a ) ) )

        if len( allowdenyList ) > 0:
            self.__allowdeny = tuple( allowdenyList )
            app.logger.info( "Restriction networks: {}".format( ", ".join( ["{a}={b}" for a,b in allowdenyList ] ) ) )  # noqa
            # Install to hook function as we have allow and/or deny networks
            app.before_request_funcs.setdefault( None, [] ).append( self._limit_remote_address )

        else:
            app.logger.warning( "No DENY or ALLOW configured in ADDRESS_RESTRICTION" )

        self.create_access_log()
        return

    def _access_log( self ) -> None:
        """ Hook function for Flask.before_request_funcs

        This function lust logs the access

        """
        self.__access.info( f"{request.remote_addr} {request.url} {request.user_agent}" )
        return

    def _limit_remote_address( self ) -> None:
        """ Hook function for Flask.before_request_funcs

        This function checks the request address against the allowed and denied addresses

        When a address is denied the Flask abort() function with 403 is called, raising the exception

        Otherowise the function just returns None

        """
        remoteAddress = ipaddress.IPv4Address( request.remote_addr )
        hit = False
        for allow, mask in self.__allowdeny:
            # Check for address match
            if remoteAddress in mask:
                # Check if its denied
                if allow == 'D':
                    self.__access.error( f"{remoteAddress} denied by rule {mask} {request.url} {request.user_agent}" )
                    abort( 403 )

                # Allowed, so exit
                self.__access.info( f"{remoteAddress} allowed by rule {mask} {request.url} {request.user_agent}" )
                hit = True
                break

        if not hit:
            self.__access.warning( f"{remoteAddress} was not in allowed/denied rules {request.url} {request.user_agent}" )

        # No, its not
        return

    @staticmethod
    def networkMask( address: str, recursive: bool = True ) -> Union[ ipaddress.IPv4Network, ipaddress.IPv6Network ]:
        """ Convert IP address, hostname or IP mask into a network address

        Returns: ipaddress.IPv4Network or ipaddress.IPv6Network

        When the address is not a valid IP address or IP network address it raises ipaddress.AddressValueError

        """
        try:
            return ipaddress.IPv4Network( address )

        except ipaddress.AddressValueError:
            pass

        except Exception:
            raise

        try:
            return ipaddress.IPv6Network( address )

        except ipaddress.AddressValueError:
            pass

        except Exception:
            raise

        if recursive:
            try:
                return FlaskAllow.networkMask( socket.gethostbyname_ex( address )[2][0], recursive = False )

            except socket.gaierror:
                raise ipaddress.AddressValueError( f"{address} is not a valid IP4, IP6 or host name")

        raise ipaddress.AddressValueError( f"{address} is not a valid IP4, IP6 or host name")
